index.js

```diff
@@ -80,17 +40,55 @@ const onError = function(context, err) {
  * @return {Promise} a Promise for whatever the wrapped function eventually
  *   resolves to.
  */
-retryRec = function(context) {
-  let result;
-
-  // Base case: last attempt
-  if (context.attempts === context.retries) {
-    result = context.fn.apply(context.fnThis, context.args);
-    return Promise.resolve(result);
-  } else {
-    // try the function. if we catch anything, wait, then retry
-    result = context.fn.apply(context.fnThis, context.args);
-    return Promise.resolve(result).catch((err) => onError(context, err));
+const execute = async function(context) {
+  /**
+   * The first attempt of executing the function shouldn't count as a "retry"
+   * since it is the initial execution. Successive executions of the function
+   * should count as actual "retries".
+   *
+   * In order to achieve this, retries is incremented at the end of each loop
+   * iteration as opposed to the beginning of each loop iteration.
+   */
+  let retries = 0;
+
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    /* eslint-disable no-await-in-loop */
+
+    try {
+      return await context.fn.apply(context.fnThis, context.args);
```

Comment:
Hm..why not just `await`? I understand that `return await` ensures that our catch block is executed if an error is thrown. Just `result` will not do that. But, do we need `return await` here? Also asking because it can seem like `return await` is not necessary and someone may just refactor it to `return`. What do you think?

Reply:
Hm http://hassansin.github.io/Why-return-await-Is-a-Bad-Idea

Reply:
The whole issue around `return await` is actually a complicated situation and I actually just changed my stance on it this past weekend, but I'll get into that in a second. 

Ignoring the discussion around `return await`, the reason we need await here is exactly for the reason you mentioned about the catch block. If we just return the promise, our catch block will not be triggered. We could refactor this into `return context.fn.apply(....).catch(function() {...`, but that will problematic for stack traces (details explained below).

Now going back to `return-await`, I had actually enabled the `no-return-await` rule in the RC version of our eslint config because of the performance issues mentioned on the eslint rule page (and in the blog post you linked). However, I'm actually going to disable that rule for three reasons:
1. The performance penalty no longer applies in Node.js 12, the V8 developers identified and patched this issue (the issue applies to all cases of `await promise` and not just `return await promise) [Faster async functions and promises - await under the hood](https://v8.dev/blog/fast-async#await-under-the-hood)
2. `await` is basically required for better stack traces [Zero-cost async stack traces](http://bit.ly/v8-zero-cost-async-stack-traces)
3. The statement `return callExpression()` is actually a potential refactoring trap as the person refactoring the code might not know that `callExpression()` actually evaluates to a promise.

Checkout the conversation on this issue thread as well: https://github.com/eslint/eslint/issues/12246